internal/controllers/machine/drain/drain_test.go

```diff
@@ -0,0 +1,968 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client/interceptor"
+)
+
+func TestRunCordonOrUncordon(t *testing.T) {
+	tests := []struct {
+		name string
+		node *corev1.Node
+	}{
+		{
+			name: "Uncordoned Node should be cordoned",
+			node: &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node-1",
+				},
+				Spec: corev1.NodeSpec{
+					Unschedulable: false,
+				},
+			},
+		},
+		{
+			name: "Cordoned Node should stay cordoned",
+			node: &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "node-1",
+				},
+				Spec: corev1.NodeSpec{
+					Unschedulable: true,
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			fakeClient := fake.NewClientBuilder().WithObjects(tt.node).Build()
+
+			drainer := &Helper{
+				Client: fakeClient,
+			}
+
+			g.Expect(drainer.CordonNode(context.Background(), tt.node)).To(Succeed())
+
+			gotNode := tt.node.DeepCopy()
+			g.Expect(fakeClient.Get(context.Background(), client.ObjectKeyFromObject(gotNode), gotNode)).To(Succeed())
+			g.Expect(gotNode.Spec.Unschedulable).To(BeTrue())
+		})
+	}
+}
+
+func TestGetPodsForEviction(t *testing.T) {
+	tests := []struct {
+		name              string
+		pods              []*corev1.Pod
+		wantPodDeleteList PodDeleteList
+	}{
+		{
+			name: "skipDeletedFilter",
+			pods: []*corev1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              "pod-1-skip-pod-old-deletionTimestamp",
+						DeletionTimestamp: &metav1.Time{Time: time.Now().Add(time.Duration(1) * time.Minute * -1)},
+						Finalizers:        []string{"block-deletion"},
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              "pod-2-delete-pod-new-deletionTimestamp",
+						DeletionTimestamp: &metav1.Time{Time: time.Now()},
+						Finalizers:        []string{"block-deletion"},
+					},
+				},
+			},
+			wantPodDeleteList: PodDeleteList{items: []PodDelete{
+				{
+					Pod: &corev1.Pod{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "pod-1-skip-pod-old-deletionTimestamp",
+						},
+					},
+					// Skip this Pod because deletionTimestamp is > SkipWaitForDeleteTimeoutSeconds (=10s) ago.
+					Status: PodDeleteStatus{
+						Delete: false,
+						Reason: PodDeleteStatusTypeSkip,
+					},
+				},
+				{
+					Pod: &corev1.Pod{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "pod-2-delete-pod-new-deletionTimestamp",
+						},
+					},
+					// Delete this Pod because deletionTimestamp is < SkipWaitForDeleteTimeoutSeconds (=10s) ago.
+					Status: PodDeleteStatus{
+						Delete:  true,
+						Reason:  PodDeleteStatusTypeWarning,
+						Message: unmanagedWarning,
+					},
+				},
+			}},
+		},
+		{
+			name: "daemonSetFilter",
+			pods: []*corev1.Pod{
+				{
```

Comment:
q: should  we add a standalone pod?

Reply:
All these test cases are running through the same filters. I already have standalone Pods in "unreplicatedFilter".

Do you think the Pods there cover everything or do you see a case that is missing there?

Reply:
It would be nice, because I saw those filters looking at ownerReferences in the Pods, which do not exists in standalone pods.

Reply:
Agree that we should test them. I'm just saying I do that already
```go
		{
			name: "unreplicatedFilter",
			pods: []*corev1.Pod{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "pod-1-delete-succeeded-pod",
					},
					Status: corev1.PodStatus{
						Phase: corev1.PodSucceeded,
					},
				},
```

(the test name is just a bit of grouping for the Pods, all tests are running through all filters)